torch/_inductor/simple_fsdp/greedy_bucket.py

```diff
@@ -0,0 +1,260 @@
+from dataclasses import dataclass, field
+from typing import Dict, List, Union
+
+import torch
+import torch.distributed as dist
+from .. import ir, scheduler
+
+from .utils import NodeType, compute_node_users, get_node_type
+from .bucket import merge_allgather, merge_ag_wait, merge_reducescatter, merge_rs_wait
+from ..comm_analysis import estimate_bucked_nccl_collective_runtime
+
+@dataclass
+class AG_INFO:
+    AG_INV_DEP: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    AG_WAIT: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    COMPUTE: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    AG_TIME: float = 0
+    COMPUTE_TIME: float = 0
+    COMPUTE_MEMORY: float = 0
+
+    # additional configs for backward
+    REDUCE_SCATTER: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    RS_WAIT: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    RS_WAIT_DEP: List["scheduler.BaseSchedulerNode"] = field(default_factory=list)
+    RS_TIME: float = 0
+
+
+def greedy_check(current_comp, current_ag, current_mem, node_ag, node_mem, memory_constraint, first_AG, current_rs=0):
+    """
+    Greedy algorithm for backward
+    """
+    if current_ag == 0:
+        return False
+
+    if first_AG:
+        return True
+
+    if current_mem + node_mem > memory_constraint:
+        return True
+
+    if current_ag + node_ag + current_rs > current_comp:
```

Comment:
this doesn't consider the effect that when merging `current_ag` and `node_ag`, you'd save a `b` in `ax + b` right?

Reply:
The current_ag is the new time, where all gathers are being bucketed. The new time estimation is performed here: https://github.com/tianyu-l/pytorch_intern24/blob/c518eb6b8a48096263ce7b73f26ef0aee1c2638b/torch/_inductor/comm_analysis.py#L350